File: leetcode/543_diameter_of_binary_tree.py
"""
T: O(N)
S: O(N)

A good example of making a recursive solution into an iterative one.
For each node, we need to get to know the depths of their left and right
branches. Using this information, we find the diameter for the current node
and current node depth. The maximum seen diameter is the maximum diameter
of the tree.
"""


from typing import Dict, Optional


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:

    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        max_diameter = 0
        if not root:
            return max_diameter
        node_depths = {None: 0}  # type: Dict[Optional[TreeNode], int]
        nodes_to_process = [root]
        while nodes_to_process:
            curr = nodes_to_process[-1]
            left, right = curr.left, curr.right
            if left and left not in node_depths:
                nodes_to_process.append(left)
            elif right and right not in node_depths:
                nodes_to_process.append(right)
            else:
                _ = nodes_to_process.pop()
                left_depth = node_depths[left]
                right_depth = node_depths[right]
                node_depths[curr] = 1 + max(left_depth, right_depth)
                max_diameter = max(max_diameter, left_depth + right_depth)
        return max_diameter

File: leetcode/java/543DiameterOfBinaryTree.java
import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

public class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;

  TreeNode(int x) {
    val = x;
  }
}

class Solution {
  public int diameterOfBinaryTree(TreeNode root) {
    Map<TreeNode, Integer> nodeDepth = new HashMap<>();
    nodeDepth.put(null, 0);
    Stack<TreeNode> nodesToProcess = new Stack<>();
    nodesToProcess.push(root);
    int maxDiameter = 0;
    while (!nodesToProcess.isEmpty()) {
      TreeNode currentNode = nodesToProcess.pop();
      if (currentNode == null) {
        continue;
      }
      TreeNode left = currentNode.left;
      TreeNode right = currentNode.right;
      if (nodeDepth.containsKey(left) && nodeDepth.containsKey(right)) {
        int leftDepth = nodeDepth.get(left);
        int rightDepth = nodeDepth.get(right);
        maxDiameter = Math.max(maxDiameter, leftDepth + rightDepth);
        nodeDepth.put(currentNode, Math.max(leftDepth, rightDepth) + 1);
      } else {
        nodesToProcess.push(currentNode);
        nodesToProcess.push(right);
        nodesToProcess.push(left);
      }
    }
    return maxDiameter;
  }
}
